File: kitten-housing-care/src/main/java/org/jesperancinha/housing/controller/CatControllerImpl.java
package org.jesperancinha.housing.controller;

import org.jesperancinha.housing.data.CatDto;
import org.jesperancinha.housing.service.CatService;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutionException;

@RestController
@RequestMapping("/cats")
public class CatControllerImpl implements CatController {

    private final CatService catService;

    public CatControllerImpl(CatService catService) {
        this.catService = catService;
    }

   public List<CatDto> getAllCats() throws ExecutionException, InterruptedException, IOException {
        return this.catService.getAllCats();
    }

   public List<CatDto> getFullAllCats() throws ExecutionException, InterruptedException {
        return this.catService.getFullAllCats();
    }

    public CatDto getCatByIdI(@PathVariable Long catId) throws IOException {
        return this.catService.getCatById(catId);
    }

    public CatDto getFullCatById(@PathVariable Long catId) throws IOException, ExecutionException, InterruptedException {
        return this.catService.getFullCatById(catId);
    }
}

File: kitten-house-care-coroutines/src/main/kotlin/org/jesperancinha/kittens/controller/CatControllerImpl.kt
package org.jesperancinha.kittens.controller

import org.jesperancinha.kittens.dto.CatDto
import org.jesperancinha.kittens.service.CatService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import java.util.concurrent.Flow

@RestController
@RequestMapping("/cats")
class CatControllerImpl(private val catService: CatService) {
    @GetMapping
    fun allCats(): kotlinx.coroutines.flow.Flow<CatDto> = catService.allCats()

    @GetMapping("/{catId}")
    suspend fun getCatByIdI(@PathVariable catId: Long)= catService.getCatById(catId)

    @GetMapping("/full/{catId}")
   suspend fun getFullCatById(@PathVariable catId: Long)= catService.getFullCatById(catId)

    @GetMapping("/full")
    fun fullAllCats(): kotlinx.coroutines.flow.Flow<CatDto> = catService.fullAllCats()

    @GetMapping("/fullForTest")
    fun fullAllCatsReactiveForTest(): kotlinx.coroutines.flow.Flow<CatDto> = catService.fullAllCats()

    @GetMapping("/full/nonreactive")
    fun fullAllCatsNonReactive(): kotlinx.coroutines.flow.Flow<CatDto> = run {
        Thread.sleep(1)
        catService.fullAllCatsNonReactive()
    }
}